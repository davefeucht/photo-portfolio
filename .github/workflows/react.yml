# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: React CI/CD

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Install Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}

    - name: Enable corepack
      run: corepack enable

    - name: Install dependencies
      run: yarn install

    - name: Build project
      run: npm run build --if-present

    - name: Run tests
      run: npm test
