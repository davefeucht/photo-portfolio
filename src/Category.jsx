/*****************
* Category component displays the posts for a particular category.
* Props:
*  id: PropTypes.number,
*  name: PropTypes.string,
*  site: PropTypes.string,
*  showAllPosts: PropTypes.bool,
*  clickCategory: PropTypes.func
*****************/

import React from "react";
import axios from "axios";
import Posts from "./Posts.jsx";
import Post from "./Post.jsx";
import PropTypes from "prop-types";

class Category extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      categoryPost: {},
      errorMsg: ""
    };
  }

  _getCategoryPost(categoryId) {
    let getCategoryPostURI = "http://" + this.props.site + "/wp-json/wp/v2/posts?categories=" + categoryId;
    axios.get(getCategoryPostURI)
      .then(res => {
        if(res.data[0] !== undefined) {
          const categoryPost = res.data[0];
          this.setState({categoryPost});
        }
      }, error => {
        const errorMsg = "Did not work: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  }

  _openCategory() {
    this.setState({showAllPosts: !this.state.showAllPosts}); 
    this.props.clickCategory(this.props.id, this.props.name);
  }

  componentWillMount() {
    this._getCategoryPost(this.props.id);
  }

  render() {
    let postBody = "";
    let categoryTitle = <div className="category-title-clickable">{this.props.name}</div>;
    let categorySubtitle = "";
    if(!this.props.showAllPosts) {
      if(!(Object.keys(this.state.categoryPost).length === 0 && this.state.categoryPost.constructor === Object)) {
        postBody = <Post key={this.state.categoryPost.id.toString()} category={this.props.id} categoryName={this.props.name} title={this.state.categoryPost.title.rendered} id={this.state.categoryPost.id} image={this.state.categoryPost.featured_media} context="category-image" site={this.props.site} />;
      }
    }
    else {
      categoryTitle = <div className="category-title">{this.props.name}</div>;
      categorySubtitle = <div className="category-subtitle">Back to Categories</div>; 
      postBody = <Posts site={this.props.site} category={this.props.id} />;
    }
    return ( 
      <div className="category" onClick={this._openCategory.bind(this)}>
        <div>
          {categoryTitle}
          {categorySubtitle}
        </div>
        {postBody}
      </div>
    );
  }

}

Category.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  site: PropTypes.string,
  showAllPosts: PropTypes.bool,
  clickCategory: PropTypes.func
};

export default Category;
