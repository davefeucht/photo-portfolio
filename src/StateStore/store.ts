import { configure, makeAutoObservable } from 'mobx';
import { ImageData, Image, Store, Category, Post, Page } from '../utils/types';

configure({ enforceActions: 'observed' });

class stateStore implements Store {
    menuState = 'closed'
    applicationRoot: HTMLElement = null;
    screenInfo: {
        width: number,
        height: number
    }
    siteInfo: {
        siteName: string,
        siteUrl: string
    }
    visiblePost: {
        postId: number,
        postTitle: string,
        tags: number[],
        tagNames: string[],
        fullImageUrl: string,
        width: number,
        height: number
    };
    categoryList: Category[] = [];
    currentCategoryPosts: Post[] = [];
    currentCategoryData: Category = {
        id: 0,
        count: 0,
        description: '',
        link: '',
        name: '',
        slug: '',
        taxonomy: '',
        parent: 0
    };
    pages: Page[] = [];
    currentPageData: Page = {
        id: 0,
        link: '',
        modified: '',
        slug: '',
        type: '',
        title: {
            rendered: ''
        },
        content: {
            rendered: '',
            protected: false
        },
        author: 0,
        featured_media: 0
    };
    maxItemsPerPage = 10;

    constructor() {
        makeAutoObservable(this);
    }

    setMenuState = (state: string) => {
        this.menuState = state;
    }

    toggleMenuState = () => {
        this.menuState = this.menuState === 'closed' ? 'open' : 'closed';
    }

    setApplicationRoot = (element: HTMLElement) => {
        this.applicationRoot = element;
    }

    setSiteName = (name: string) => {
        this.siteInfo.siteName = name;
    }

    setVisiblePost = (postId: number, postTitle: string) => {
        this.visiblePost.postId = postId;
        this.visiblePost.postTitle = postTitle;
    }

    setVisiblePostImage = (fullImageUrl: string) => {
        this.visiblePost.fullImageUrl = fullImageUrl;
    }

    setVisiblePostTags = (tagNames: string[]) => {
        tagNames.forEach(tagName => {
            this.visiblePost.tagNames.push(tagName);
        });
    }

    clearVisiblePostTagNames = () => {
        this.visiblePost.tagNames.splice(0, this.visiblePost.tagNames.length);
    }

    setCategoryList = (categories: Category[]) => {
        this.categoryList.length = 0;
        categories.forEach((category, index) => {
            this.categoryList[index] = category;
        })
    }

    setCategoryPosts = (posts: Post[]) => {
        this.currentCategoryPosts.length = 0;
        posts.forEach((post, index) => {
            this.currentCategoryPosts[index] = post;
        })
    }

    setCategoryData = (categoryData: Category) => {
        Object.keys(categoryData).forEach((key: keyof Category) => {
            this.currentCategoryData[key] = categoryData[key];
        });
    }

    setThumbnailImageUrl = (imageData: ImageData) => {
        this.currentCategoryPosts[imageData.post_index].thumbnail_image = imageData.image_url;
    }

    setCurrentPost = (postData: Post) => {
        Object.keys(postData).forEach((key: keyof Image) => {
            this.visiblePost[key] = postData[key];
        })
    }

    setPages = (pages: Page[]) => {
        pages.forEach(page => {
            this.pages.push(page);
        });
    }

    setPageData = (pageData: Page) => {
        Object.assign(this.currentPageData, pageData);
    }
}

export default stateStore;
