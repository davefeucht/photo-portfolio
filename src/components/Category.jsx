/*****************
* Category component displays the posts for a particular category.
* Props:
*  id: PropTypes.number,
*  name: PropTypes.string,
*  site: PropTypes.string,
*  clickCategory: PropTypes.func
*****************/

import React from "react";
import axios from "axios";
import Posts from "./Posts.jsx";
import CategoryHeader from "./CategoryHeader.jsx";
import StyledCategory from "./styledComponents/StyledCategory.jsx";
import PropTypes from "prop-types";

export default class Category extends React.Component {

  static propTypes = {
    id: PropTypes.number,
    name: PropTypes.string,
    site: PropTypes.string,
    clickCategory: PropTypes.func
  };

  state = {
    categoryPost: {},
    errorMsg: ""
  };

  //Function to get the 'main' post for the category
  _getCategoryPost(categoryId) {
    let getCategoryPostURI = "http://" + this.props.site + "/wp-json/wp/v2/posts?categories=" + categoryId;
    axios.get(getCategoryPostURI)
      .then(res => {
        if(res.data[0] !== undefined) {
          const categoryPost = res.data[0];
          this.setState({categoryPost});
        }
      })
      .catch(error => {
        const errorMsg = "Could not get posts for this category: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  }

  //When the component is about to mount, get the main post for the category
  componentWillMount() {
    this._getCategoryPost(this.props.id);
  }

  render() {
    return ( 
      <StyledCategory>
        <CategoryHeader id={this.props.id} name={this.props.name} clickCategory={this.props.clickCategory.bind(this)} />
        <Posts site={this.props.site} category={this.props.id} />
      </StyledCategory>
    );
  }
}