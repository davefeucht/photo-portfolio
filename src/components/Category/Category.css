.category {
  min-height: 100%;
  display: grid;
  grid-template-columns: 100%;
  grid-template-areas: "category-header" "category-content";
  grid-template-rows: .25fr 4fr;
}