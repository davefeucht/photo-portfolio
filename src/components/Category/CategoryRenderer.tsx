import Posts from 'components/Posts/Posts';
import SectionHeader from 'components/SectionHeader/SectionHeader';
import { observer } from 'mobx-react';
import * as React from 'react';
import { Outlet } from 'react-router-dom';
import { Post, ScreenInfo } from 'utils/types';

interface Props {
    categoryId: number;
    maxItemsPerPage: number;
    categoryName?: string;
    currentCategoryPosts: Post[];
    screenInfo: ScreenInfo;
}

const CategoryRenderer: React.FC<Props> = ({
    categoryId,
    maxItemsPerPage,
    categoryName,
    currentCategoryPosts,
    screenInfo
}) => {
    return (
        <div className="category">
            <SectionHeader title={categoryName ?? ""} />
            <Posts
                maxItemsPerPage={maxItemsPerPage}
                screenInfo={screenInfo}
                categoryId={categoryId}
                currentCategoryPosts={currentCategoryPosts}
            />
            <Outlet />
        </div>
    );
};

export default observer(CategoryRenderer);
