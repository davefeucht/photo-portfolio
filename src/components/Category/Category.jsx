/*****************
* Category component displays the posts for a particular category.
*****************/

import React from 'react';
import { observer } from 'mobx-react';
import { 
  useParams,
} from 'react-router-dom';
import Posts from '../Posts/Posts.jsx';
import CategoryHeader from '../CategoryHeader/CategoryHeader.jsx';
import './Category.css';

const Category = observer(({ stateStore, api }) => {
  let { categoryId } = useParams();

  api.getPosts(categoryId);
  api.getCategoryInfo(categoryId);

  return ( 
    <div className="category">
      <CategoryHeader stateStore={stateStore} categoryData={stateStore.currentCategoryData}/>
      <Posts stateStore={stateStore} categoryId={categoryId} api={api} />
    </div>
  );
});

Category.displayName = 'Category';

export default Category;