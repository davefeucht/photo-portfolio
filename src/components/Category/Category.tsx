/** ***************
* Category component displays the posts for a particular category.
**************** */

import './Category.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect } from 'react';
import {
    Outlet,
    useParams
} from 'react-router-dom';

import { getCategoryInfo, getPosts, getPostThumbnail } from '../../utils/Api';
import {
    Category as CategoryState,
    ImageData,
    Post as PostState,
    ScreenInfo,
    SiteInfo
} from '../../utils/types';
import Posts from '../Posts/Posts';
import SectionHeader from '../SectionHeader/SectionHeader';

interface CategoryProps {
    maxItemsPerPage: number,
    siteInfo: SiteInfo,
    screenInfo: ScreenInfo,
    currentCategoryPosts: PostState[],
    currentCategoryData: CategoryState,
    setCategoryPosts: (a: PostState[]) => void,
    setCategoryData: (a: CategoryState) => void,
    setThumbnailImageUrl: (a: ImageData) => void
}

const Category: React.FC<CategoryProps> = ({
    maxItemsPerPage,
    siteInfo,
    screenInfo,
    currentCategoryPosts,
    currentCategoryData,
    setCategoryPosts,
    setCategoryData,
    setThumbnailImageUrl
}) => {
    const { categoryId = '' } = useParams();

    useEffect(() => {
        if (categoryId) {
            getPosts(parseInt(categoryId), siteInfo.siteUrl)
                .then((posts: PostState[]) => {
                    setCategoryPosts(posts);
                    currentCategoryPosts.forEach((post, index) => {
                        getPostThumbnail(post.featured_media, siteInfo.siteUrl)
                            .then((thumbUrl: string) => {
                                setThumbnailImageUrl({ post_index: index, image_url: thumbUrl });
                            });
                    });
                });
            getCategoryInfo(parseInt(categoryId), siteInfo.siteUrl)
                .then((categoryInfo: CategoryState) => {
                    setCategoryData(categoryInfo);
                });
        }
    }, [categoryId]);

    return (
        <div className="category">
            <SectionHeader title={currentCategoryData.name} />
            <Posts
                maxItemsPerPage={maxItemsPerPage}
                screenInfo={screenInfo}
                categoryId={parseInt(categoryId)}
                currentCategoryPosts={currentCategoryPosts}
            />
            <Outlet />
        </div>
    );
};

Category.displayName = 'Category';

export default observer(Category);
