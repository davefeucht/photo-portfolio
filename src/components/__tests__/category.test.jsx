import { render } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';

import StateStore from '../../StateStore/store';
import Category from '../Category/Category';

const posts = [
    {
        id: 887,
        featured_media: 142,
        categories: [220]
    },
    {
        id: 824,
        featured_media: 820,
        categories: [220]
    }
];

const category = {
    id: 35,
    count: 3,
    description: 'Category of things',
    link: '',
    name: 'Some category',
    slug: 'smcg',
    taxonomy: '',
    parent: 23,
    meta: {}
};

jest.mock('../../utils/Api', () => ({
    getPosts: () => Promise.resolve(posts),
    getPostThumbnail: () => Promise.resolve('https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg'),
    getCategoryInfo: () => Promise.resolve(category),
    getCategoryImage: () => Promise.resolve('https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg')
}));

test('Category displays', async () => {
    const store = new StateStore();
    let container;
    await act(async () => {
        container = render(
            <MemoryRouter initialEntries={['/category/35']}>
                <Category stateStore={store} />
            </MemoryRouter>
        ).container;
    });
    expect(container.firstChild).toMatchSnapshot();
});
