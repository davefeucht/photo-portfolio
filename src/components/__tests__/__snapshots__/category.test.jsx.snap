// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category displays 1`] = `
<div
  class="category"
>
  <div
    class="section-header"
  >
    <div
      class="title"
    />
  </div>
  <div
    class="posts"
  >
    <div
      class="pagination-navigation"
    >
      <div
        class="pagination-navigation__content"
      />
    </div>
  </div>
</div>
`;
